web3-functions/kill-switch/index.ts

```diff
@@ -0,0 +1,72 @@
+import { Contract } from '@ethersproject/contracts'
+import { Web3Function, Web3FunctionContext } from '@gelatonetwork/web3-functions-sdk'
+
+import { killSwitchOracleAbi, multicallAbi, oracleAbi } from '../../abis'
+import { addresses } from '../../utils'
+
+Web3Function.onRun(async (context: Web3FunctionContext) => {
+    const { multiChainProvider } = context
+
+    const provider = multiChainProvider.default()
+
+    const multicall = new Contract(addresses.mainnet.multicall, multicallAbi, provider)
+    const killSwitchOracle = new Contract(addresses.mainnet.killSwitchOracle, killSwitchOracleAbi, provider)
+
+    const oracleAddresses = await killSwitchOracle.oracles()
+
+    let multicallCalls: Array<{ target: string; callData: string }> = []
+
+    for (const oracleAddress of oracleAddresses) {
+        const oracle = new Contract(oracleAddress, oracleAbi, provider)
+        multicallCalls = [
+            ...multicallCalls,
+            ...[
+                {
+                    target: oracleAddress,
+                    callData: oracle.interface.encodeFunctionData('latestAnswer'),
+                },
+                {
+                    target: addresses.mainnet.killSwitchOracle,
+                    callData: killSwitchOracle.interface.encodeFunctionData('oracleThresholds', [oracleAddress]),
+                },
+                {
+                    target: addresses.mainnet.killSwitchOracle,
+                    callData: killSwitchOracle.interface.encodeFunctionData('triggered'),
+                },
+            ],
+        ]
+    }
+
+    let multicallResults = (await multicall.callStatic.aggregate(multicallCalls)).returnData
+
+    for (const oracleAddress of oracleAddresses) {
+        const oracle = new Contract(oracleAddress, oracleAbi, provider)
+
+        const latestAnswer = BigInt(
+            oracle.interface.decodeFunctionResult('latestAnswer', multicallResults[0])[0].toString(),
+        )
+        const threshold = BigInt(
+            killSwitchOracle.interface.decodeFunctionResult('oracleThresholds', multicallResults[1])[0].toString(),
+        )
+        const triggered = killSwitchOracle.interface.decodeFunctionResult('triggered', multicallResults[2])[0]
+
+        multicallResults = multicallResults.slice(3)
+
+        if (!triggered && latestAnswer > 0 && latestAnswer <= threshold) {
```

Comment:
We can take out the triggered bool here cause the contract doesn't check that

Reply:
Once triggered, it can still be triggered again to lock, though if we don't check this, we'll be calling trigger every block, when one of the prices is below threshold, even if the trigger was already called and further calls will make no effect. Situations when another call will be needed will rather rare (far more rare than triggering kill switch, which itself will be rare), so I'd rather leave it as is.
Maybe @hexonaut has an opinion here?

Reply:
Yes it should only automatically trigger the first time. Being able to trigger it multiple times is only in the edge case where there is a spell to add a new collateral under a timelock in the GSM such that when it activates we can disable the borrow on the new collateral right away due to the emergency.

Reply:
Ah yeah of course makes sense